### PR TITLE
Efficient labels mapping for drawing in Labels (60 FPS even with 8000x8000 images)

### PR DESCRIPTION
**Pull request by ksofiyuk**
# Description

This PR introduces two independent optimizations for labels rendering during drawing: local updates and caching optimizations.

These optimizations virtually solve any performance issues related to drawing in the Labels layer. With them, it is possible to achieve 60 FPS when drawing even in 8000x8000 images.

### Local updates

The initial implementation updates the whole labels map and copies it to the shader on each brush update, which becomes a major bottleneck when you work with high-resolution images no matter how effective the labels mapping implementation is. 

In this PR, only partial updates are sent to the VisPy labels layer. Each time `Layers.data_setitem` is called, it tracks the bounds of modified region, and instead of calling `Labels.refresh()` that triggers the update of the whole labels image, it calls `Labels._partial_labels_refresh` that emits the `Labels.events.labels_update` event that comes with the slice localizing the modified region, which is then handled by `VispyLabelsLayer._on_partial_labels_update`. VisPy textures can be partially updated, which is used in the `_on_partial_labels_update` method.

### Caching optimization

The idea is to recompute color mapping only for the elements of a label map that are changed from a previous update. Even in the parts of the code where the color mapping is implemented quite efficiently it can give up to 5x speedup by avoiding slow `np.float32` recomputations for most pixels.

In practice, when you use brush, less than 1% of pixels are updated at each iteration (on large images the percentage is even smaller). As a result, this optimization should work and give a significant boost in 99%+ of typical use case scenarios.

## Benchmark results

I added a new benchmark (benchmark_labels_layer.LabelsDrawing2DSuite) that measures the timings of brush drawing in different modes (auto/direct, contour == 0/1) with different brush sizes. It simulates the brush drawing from the position (0, 0) to (n - 1, n - 1) with 30 refresh updates along the way.

This PR:
```                                                                                                                          
              ====== ============ ============ ============ ============ ============
              --                                  color_mode / contour
              ------------------- ---------------------------------------------------
                n     brush_size    auto / 0     auto / 1    direct / 0   direct / 1
              ====== ============ ============ ============ ============ ============
               512        8        28.2±0.3ms    33.6±1ms     29.2±1ms    32.9±0.7ms
               512        64       22.8±0.1ms   28.1±0.2ms   22.9±0.1ms   29.1±0.3ms
               512       256       128±0.9ms    153±0.7ms    128±0.7ms     153±1ms
               3072       8        147±0.6ms    165±0.7ms     147±3ms     166±0.4ms
               3072       64       131±0.9ms     151±2ms     132±0.8ms     150±1ms
               3072      256        450±2ms      502±2ms     452±0.7ms     501±2ms
              ====== ============ ============ ============ ============ ============
```

main 7b3f7ecfe0e74af5f52dd3eb380b5752b2f3378c:
```
              ====== ============ ============ ============ ============ ============
              --                                  color_mode / contour
              ------------------- ---------------------------------------------------
                n     brush_size    auto / 0     auto / 1    direct / 0   direct / 1
              ====== ============ ============ ============ ============ ============
               512        8        86.6±0.7ms    242±2ms      100±1ms      256±1ms
               512        64        75.6±1ms     230±1ms      90.6±1ms     241±1ms
               512       256       163±0.9ms     316±2ms      180±1ms      328±1ms
               3072       8        1.20±0.01s   6.33±0.02s   1.80±0.02s   6.90±0.02s
               3072       64       1.15±0.01s    6.30±0s      1.77±0s      6.90±0s
               3072      256       1.30±0.01s   6.40±0.01s    1.92±0s     7.04±0.02s
              ====== ============ ============ ============ ============ ============
```

The difference is most noticeable when the heavy labels mapping code is used (contour=1), with small brush sizes it gives almost 50x performance boost.

Please keep in mind that this benchmark only measures the time it takes to update `layer.data` + the time it takes to convert labels to colors before sending them to VisPy. This benchmark does not take into account the time it takes to transfer data from napari to OpenGL or the time it takes to process labels in any type of OpenGL shader. And the local updates also give a substantial speed up in the part that is not measured.

## Type of change

- [X] Optimization (non-breaking change which speedups existing code)

# How has this been tested?
- [X] all tests pass with my change

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/layers/labels/labels.py
Comment (alisterburt) [inline napari/layers/labels/labels.py]:
```suggestion

        # cache the raw data and keep track of when values are changed
        changed_mask = None
```

**Review by alisterburt — APPROVED** [review]
very nice work @ksofiyuk !

Changes look great to me, typing test failure seems unrelated (as in #5723) - for me this is ready to merge

supersedes #5723

**Comment by ksofiyuk** [comment]
@alisterburt Actually, there is a small problem. It failed some tests for the Label layer. It turned out that this optimization cannot be easily applied when `Layer.contour > 0` as morphological operations are performed on the `raw` array for computing contours. I'll disable it when `Layer.contour > 0`.

File: napari/layers/labels/labels.py
Comment (alisterburt) [inline napari/layers/labels/labels.py]:
```suggestion
        if self.contour == 0:  # morphological operations require whole image
```

**Comment by alisterburt** [comment]
ooh, I just saw you brought it back, could you explain how this is working?

**Comment by ksofiyuk** [comment]
I found out that there is no need to disable it when `contour > 0`, it is enough just to move the cache initialization after the if block, in which contours are computed. However, computing contours at each update makes caching much less useful, but it should give some gain anyway.

**Comment by Czaki** [comment]
@alisterburt Labels and milestone (are you sure for 0.4)?

**Review by jni — CHANGES_REQUESTED** [review]
Couple of comments here,

1) #5723 seems to be reasonably independent of this — could we merge that first? It will make the changes more granular and easier to review — #5723 was already ready to merge so I kinda wanna get that in while we review this. Does that work @ksofiyuk? They look to me like they sort of touch different parts of the file. But,

2) Another reason why I want to have the two reviews separately is that I think there should be a bit more of a change in the middle to make it clearer what's going on. For example, I don't like reusing variable names (`raw_modified = raw_modified[changed_mask]`), as it makes it harder to reason about the code (is this image shaped or a linear array of changed values?). I also don't like that now, in most cases, `image = ` will actually *not* be an image, but a linear set of values. So I'd like a bit of an update to the logic and variable names to better reflect what's going on.

3) Finally, while #5723 has no bearing on #3308, this *does* have a bearing on it — it looks like caching will be beneficial to #3308 also, except we are caching `.astype(np.float32)` instead of the complicated legacy logic here. So having a separate PR will make it easier to port those relevant changes to #3308.

btw, I'm super curious — I thought that this would actually track changed pixels from painting and so on, in which case the speedup would be obvious. But no, it actually *checks* for changes by doing a full `array != value`, which is *not* immediately obviously faster than doing `array.astype()`. But your benchmarks (and my own experimentation with this PR 🙏) suggest that it definitely is faster! Do I understand correctly that that's what you're doing here?

**Comment by jni** [comment]
I just saw that indeed this simply has all of the original commits from #5723, so it will be very straightforward to rebase/merge main after that merges. I'm gonna go ahead and do that.

**Comment by jni** [comment]
Oh! One more Q: could you please test this solution with zarr, tensorstore, or dask arrays as labels? I'm concerned about some of these operations (boolean indexing) in those contexts...

**Comment by jni** [comment]
@ksofiyuk are you comfortable with cherry-picking (should be easier than merging here)? If not I can do it for you and force push.

**Comment by ksofiyuk** [comment]
> Another reason why I want to have the two reviews separately is that I think there should be a bit more of a change in the middle to make it clearer what's going on.

@jni No problem at all, I assumed to have them reviewed separately that's why I created a separate pull request. 

>  For example, I don't like reusing variable names (raw_modified = raw_modified[changed_mask]), as it makes it harder to reason about the code (is this image shaped or a linear array of changed values?). I also don't like that now, in most cases, image =  will actually not be an image, but a linear set of values. So I'd like a bit of an update to the logic and variable names to better reflect what's going on.

I agree on the part on the `image` naming, it's confusing. Maybe it's better to use something like `mapped_labels` instead? This name doesn't have any assumptions about the dimensionality of an array and I think it describes the variable better.

If we want to refactor the variables names in this method, then I suggest changing `raw` to `labels` as I think it describes the object better. `raw_modified` to `labels_modified` or `labels_changed`. I don't fully agree about a remark on reusing `raw_modified` as this PR actually makes this initial name more appropriate in this method. As you can interpret it as a part of `raw` that is modified and needs to be recalculated. In the initial implementation the name `raw_modified` was confusing for me. Why was it called `modified` if it was equal to `raw` (most common scenario), except for the case when contours are used?

> I thought that this would actually track changed pixels from painting and so on, in which case the speedup would be obvious. But no, it actually checks for changes by doing a full array != value, which is not immediately obviously faster than doing array.astype(). But your benchmarks (and my own experimentation with this PR pray) suggest that it definitely is faster! Do I understand correctly that that's what you're doing here?

You understood everything absolutely correctly. I implemented this caching only to avoid recomputation of the label mapping in floats. So it replaces several slow float operations on a full array with just a single comparison of integers and taking a subarray using a boolean mask. That's where the performance boost comes from.

> I thought that this would actually track changed pixels from painting and so on, in which case the speedup would be obvious

I'm still thinking that it is a good thing to do, but it requires a deep refactoring not only of the Label layer, but also changing the way how the labels are sent to GPU. In the current implementation, optimizing `_raw_to_displayed` further just doesn't make any sense as the whole array is sent to VisPy which performs a lot of float operations with this array before even sending it to GPU, which is now a bottleneck.

I think it's better to do it based on #3308.

> could you please test this solution with zarr, tensorstore, or dask arrays as labels? I'm concerned about some of these operations (boolean indexing) in those contexts...

I assume they should follow the numpy indexing convention, otherwise there are a lot of places where they can be broken. For example, in the contours computation code:
https://github.com/napari/napari/blob/5e8ffc6c1bd5d8dcd3b507b64a84cb4c5d1a2c1d/napari/layers/labels/labels.py#L905-L913

Btw, does napari have any examples of using these arrays as labels?

**Comment by ksofiyuk** [comment]
>  But no, it actually checks for changes by doing a full array != value, which is not immediately obviously faster than doing array.astype(). But your benchmarks (and my own experimentation with this PR pray) suggest that it definitely is faster! Do I understand correctly that that's what you're doing here?

If you want to have a good performance on CPU, you need to get rid of any float operations as much as possible :smile:

**Comment by ksofiyuk** [comment]
> are you comfortable with cherry-picking (should be easier than merging here)? If not I can do it for you and force push.

I'm not fully understand why cherry-picking is necessary here, but if you can do it yourself, I don't mind at all :smile:

**Comment by Czaki** [comment]
I think that they think about resolve merge conflicts.

**Comment by jni** [comment]
> I agree on the part on the image naming, it's confusing. Maybe it's better to use something like mapped_labels instead? This name doesn't have any assumptions about the dimensionality of an array and I think it describes the variable better.

I like this suggestion a lot. 😊 

> I don't fully agree about a remark on reusing raw_modified as this PR actually makes this initial name more appropriate in this method. As you can interpret it as a part of raw that is modified and needs to be recalculated.

First I want to say that we don't necessarily stand by our prior variable names, sometimes it was just "good enough" to get it in at the time and it's definitely up to revisit.

But, on this point specifically, I want to push back: in the implementation before this PR, `raw_modified` always points to the same array, although that array gets modified in-place in some code paths. But it never points to a *different* variable which is a different array with a *totally different shape*, as `raw_modified = raw_modified[changed_mask]` does.

> If we want to refactor the variables names in this method, then I suggest changing raw to labels as I think it describes the object better. raw_modified to labels_modified or labels_changed. [...] In the initial implementation the name raw_modified was confusing for me. Why was it called modified if it was equal to raw (most common scenario), except for the case when contours are used?

I think here we wanted to harmonize the code paths, so that it would always be the same variable in the end that got mapped. But we are definitely not wedded to that flow and those names and please feel free to refactor in a way that is less confusing. (It might help to make some helper functions e.g. `_find_contours`.) (It might also make my job easier when integrating with #3308! 😅)

I think the *input* name should stay as `raw` because changing it would affect other classes — I think this is overriding a base class method, so it will be a headache to update everything — the reason it's not labels is that this function is in the base image class as well. So, for the sake of simplicity, don't change the input, but feel free to rename it in the first line, e.g.

```python
labels = raw  # for readability
```

> So it replaces several slow float operations on a full array with just a single comparison of integers and taking a subarray using a boolean mask. That's where the performance boost comes from.

Replacing the rather expensive float operations makes sense. But faster than .astype(float), I guess "If you want to have a good performance on CPU, you need to get rid of any float operations as much as possible 😄" is exactly correct! I wasn't quite aware of the magnitude of this! 😅 

> I'm still thinking that it is a good thing to do, but it requires a deep refactoring not only of the Label layer, but also changing the way how the labels are sent to GPU. [...] I think it's better to do it based on https://github.com/napari/napari/pull/3308.

💯 yeah we should definitely get this in before going down that road. It actually requires *deep* rethinking of many things because plugins can also modify the `.data` array without using any of our paint methods. See e.g. #3155 for an early attempt to deal with such issues. So, it's a big can of worms, I agree that it's worth doing at some point but definitely not in this PR! 😅 

> I assume they should follow the numpy indexing convention, otherwise there are a lot of places where they can be broken.

True, but we've found a lot of such edge cases. I don't know the exact path to examples (if we have any in the gallery, you can definitely find some in the tests with `git grep {zarr,tensorstore,dask}`), but I just played around with them and everything seems fine! All of the libraries have easy creation from NumPy arrays so you can easily test their APIs:

```python
import numpy as np
import zarr
import dask.array as da
import tensorstore as ts
from skimage import data
import napari

cells = data.cells3d()
nuclei = cells[:, 1]

viewer, layers = napari.imshow(cells, channel_axis=1)

labels_np = np.zeros_like(nuclei, dtype=np.int64)
labels_zarr = zarr.array(labels_np, chunks=(20, 100, 100))
labels_dask = da.from_zarr(labels_zarr)
labels_tensorstore = ts.array(labels_np)

viewer.add_labels(labels_zarr)
viewer.add_labels(labels_dask)
viewer.add_labels(labels_tensorstore)

napari.run()
```

I tested this and other than **serious** performance issues with zarr (!!!), which are also in main anyway, all three are working. 🚀 (Can't paint into dask, and the dask cache means I can't see changes even after painting into the zarr, but that's all expected.)

Ok so now that the commits are rebased on the latest main (and indeed the changes are much smaller and easier to review! 🥳), do you mind addressing the name/flow refactoring discussed above? 🙏 

Thanks!

**Comment by ksofiyuk** [comment]
@jni I refactored the names and some flow of the method. Please take a look and let me know if it works for you.

**Comment by Czaki** [comment]
I just found that `np.unique` is pretty slow. 
I start playing with my own implementation (on my machine speedup is ~10x)
https://github.com/4DNucleome/PartSegCore-compiled-backend/blob/fast_unique/notebooks/performance.ipynb

Maybe some custom unique implementation may be a good idea as the next step for speedup labels with big numbers. 

The whole code is here https://github.com/4DNucleome/PartSegCore-compiled-backend/pull/3/files

If someone will check if I do not do some bugs or found inspiration it may be nice.

**Comment by ksofiyuk** [comment]
I noticed that `_raw_to_displayed` returns mapped colors in `float64` when the auto color mode is used. VisPy does not support `float64` and converts it all to `float32` before sending it to GPU, which leads to the unnecessary conversion overhead.

This simple change provides +30% FPS boost  when the brush is used on large images (On a `5000x5000` image the FPS goes from 6 to 8 on my machine)
https://github.com/napari/napari/blob/30d9fb546392559e43f984e06f4ec076a3a87d14/napari/utils/colormaps/colormap_utils.py#L251-L252

**Comment by ksofiyuk** [comment]
@jni I also refactored the structure of `_raw_to_displayed` by moving some code to 2 new functions `_labels_utils.get_contours` and `Label._map_labels_to_colors`. I think the original method became too long and heavy, and at least the contours computation can be moved outside.

**Comment by jni** [comment]
> I also refactored the structure of _raw_to_displayed by moving some code to 2 new functions

Great stuff! I was actually in the process of doing just that myself because indeed it was too long and I was finding it hard to even review. Breaking it up is exactly the right approach! I'll play with it a bit more today! 😊

**Comment by ksofiyuk** [comment]
I found and fixed a bug that occasionally resulted in erroneous color mapping when switching between the auto and direct color modes during masks editing. The solution is to invalidate the cache when the color mode is changed.

**Comment by alisterburt** [comment]
@jni now that #5723 has been merged I think your major points have been addressed

> https://github.com/napari/napari/pull/5723 seems to be reasonably independent of this — could we merge that first? It will make the changes more granular and easier to review — https://github.com/napari/napari/pull/5723 was already ready to merge so I kinda wanna get that in while we review this. Does that work @ksofiyuk? They look to me like they sort of touch different parts of the file. But,

> Another reason why I want to have the two reviews separately is that I think there should be a bit more of a change in the middle to make it clearer what's going on. For example, I don't like reusing variable names (raw_modified = raw_modified[changed_mask]), as it makes it harder to reason about the code (is this image shaped or a linear array of changed values?). I also don't like that now, in most cases, image =  will actually not be an image, but a linear set of values. So I'd like a bit of an update to the logic and variable names to better reflect what's going on.

> Finally, while https://github.com/napari/napari/pull/5723 has no bearing on https://github.com/napari/napari/pull/3308, this does have a bearing on it — it looks like caching will be beneficial to https://github.com/napari/napari/pull/3308 also, except we are caching .astype(np.float32) instead of the complicated legacy logic here. So having a separate PR will make it easier to port those relevant changes to https://github.com/napari/napari/pull/3308.

Now that different array types have been manually tested do we want tests in the test suite for those or are we good to go here? 🙂

**Comment by jni** [comment]
Oh yeah that's addressed! 😊 I just want to play with refactoring this function a bit more — sorry, this week turned out to be crazy but it's at the top of my to do list, will get to it shortly, probably Monday, Wednesday at the latest! 🙏

**Comment by alisterburt** [comment]
No problem, hope you're doing okay! ☺️

**Comment by ksofiyuk** [comment]
@jni @alisterburt @brisvag I've finally implemented the local updates optimization when you draw in the Labels layer (only updated portions of the array are sent to the GPU instead of the full array). It turned out to be lot easier than I was expecting. I implemented it into this PR (please take a look at my latest commit b6bb3ac).

It completely solves any performance issues related to drawing in the Labels layer. With the caching optimization and local updates, I have 60 FPS when I draw even in a 8000x8000 image.

This PR has been sitting here for several weeks, completely ready (in my opinion) to be merged, I use it internally all the time without any issues. I think it is important to merge these optimization as they make the Labels layer usable on high-resolution images. With the local updates implementation, you can use it even with extremely large images almost without any performance degradation. If you have any concerns regarding the code, please let me know.

**Comment by Czaki** [comment]
I was paling with implementing faster unique for labels like data, but struggled with compilation on MacOS (as clang is handicapped, but only default one). 

But during work, @brisvag pointed out that `pd.unique` works faster than np.unique. You may see compression here: https://github.com/4DNucleome/PartSegCore-compiled-backend/blob/eb1f704064cb508ddef96883cbdb946bfb50302a/notebooks/performance.ipynb

So maybe for big numbers, it will be better to use `pd.unique` (the result is unordered).

**Comment by ksofiyuk** [comment]
@Czaki I don't think that we need to optimize `np.unique`. In 99.9% of practical use cases, the range of labels can be localized inside the 65536 range, and the `min/max` implementation covers it. In the remaining 0.1% of situations, the caching optimization actually eliminates any performance issues even if we use ineffective `np.unique` as the size of the `labels_to_map` array rarely exceeds several thousand pixels in total. You can easily check it by setting the range from `65536` to `0` in `if upper_bound_n_unique_labels < 65536` in this PR, and you will see no performance issues even when you use very slow `np.unique` in the direct color mode.

**Comment by ksofiyuk** [comment]
The coolest thing is that the performance boosts from the local updates and caching optimizations are not just summed up, but they are multiplied and reinforce each other.

@jni The changes are tracked and localized now, so `_raw_to_displayed` compares only sliced subarrays around changes to compute `update_mask`. As a result, your initial assumption is now correct 😄 :
>  I thought that this would actually track changed pixels from painting and so on, in which case the speedup would be obvious.

File: napari/layers/labels/labels.py
Comment (alisterburt) [inline napari/layers/labels/labels.py]:
```suggestion
        # self.events.data(value=self.data)
```

Could you explain why this was commented out?

File: napari/layers/labels/labels.py
Comment (alisterburt) [inline napari/layers/labels/labels.py]:
```suggestion
        # If there are no changes, just return the cached image
```

**Review by alisterburt — APPROVED** [review]
bravo @ksofiyuk - this is an awesome PR, the performance gains are insane and the code comparatively simple ❤️ 

I've previously approved but the scope of the PR has changed quite significantly and the requested changes from @jni have not yet been fully addressed so I would want confirmation from him before dismissing his review and merging. He is away right now although I'm not sure for how long - @Czaki do you know? I'm sorry to slow you down here - we will make sure this is in for the next napari release :-)

In particular, I notice a discussion about different array backends which seems unresolved for now?

File: napari/layers/labels/labels.py
Comment (ksofiyuk) [reply napari/layers/labels/labels.py]: Thank you for pointing out! It is just a typo, I commented it accidentally. I fixed it.

**Comment by ksofiyuk** [comment]
> In particular, I notice a discussion about different array backends which seems unresolved for now?

As far as I understand, this one is resolved. @jni had some concerns that other array backends can be broken by the boolean mask indexing, but he tested it and it worked well:
> I don't know the exact path to examples (if we have any in the gallery, you can definitely find some in the tests with git grep {zarr,tensorstore,dask}), but I just played around with them and everything seems fine!
> ...
> I tested this and other than serious performance issues with zarr (!!!), which are also in main anyway, all three are working. 🚀 (Can't paint into dask, and the dask cache means I can't see changes even after painting into the zarr, but that's all expected)

Actually, using anything other than the numpy arrays in `_raw_to_displayed` doesn't make any sense since the function returns a numpy array of the same size as the input array no matter what array backend is used. We might as well just do something like that at the beginning `raw = np.array(raw)` and it should not affect anything.

**Comment by jni** [comment]
@ksofiyuk going over this now. I'm super impressed with the performance you've managed to squeeze out of the CPU-only code 😍, but I'd like some additional tests to make sure things are working and *stay* working as we continue to make edits e.g. when integrating this work into #3308. Typically, whenever I encounter a bug after some time playing with a thing (e.g. commits https://github.com/napari/napari/commit/68e8d5da8fa867e201b6fd9b06c1c0e246ff9365 and https://github.com/napari/napari/commit/8f446da411c7513df6f4727e93a415b70ada4fa0), I like to add a test to make sure the bug doesn't come up again. Cache invalidation is one of the trickiest things to get right, and I definitely don't trust myself to do it in #3308 in the absence of tests here!

Another thing I'd like explicit testing for, now that you are tracking painted pixels, is that `labels_layer.data[i, j] += 1; labels_layer.refresh()` updates things correctly.

> As far as I understand, this one is resolved. @jni had some concerns that other array backends can be broken by the boolean mask indexing, but he tested it and it worked well:

👍 correct

> Actually, using anything other than the numpy arrays in _raw_to_displayed doesn't make any sense since the function returns a numpy array of the same size as the input array no matter what array backend is used. We might as well just do something like that at the beginning `raw = np.array(raw)` and it should not affect anything.

Yeah come to think of it, by the time we get to that function we probably have indeed already instantiated a NumPy array using slicing...

**Comment by jni** [comment]
btw @ksofiyuk could you please update the benchmarks in the top-level description to match the current performance? That description is what goes into the commit message when the PR gets merged so it should be kept accurate. 🙏

**Comment by jni** [comment]
Wow I have to say I love the implementation of the local updates. 😍

I might suggest making the `updated_slice` attribute an actually valid slice, ie when you build it use `slice(min(axis_indices), max(axis_indices)) for ...`, and when you use it use `offset = sl.start for sl in updated_slice`? Personally I dislike lists of tuples of numbers or tuples of tuples of numbers because they become difficult to interpret in the case that ndim=2. But a tuple of slices is very straightforward to interpret.

What do you think?

**Comment by ksofiyuk** [comment]
I also implemented local updates to the contours visualization (`layer.contour > 0`) in the latest commits, which makes it quite responsive now. Previously they were incredibly slow when you draw even with not very large images.

> I might suggest making the updated_slice attribute an actually valid slice, ie when you build it use slice(min(axis_indices), max(axis_indices)) for ..., and when you use it use offset = sl.start for sl in updated_slice?

@jni Fully agreed, it is much simpler to expect them in the standard slicing format everywhere. I changed it in the latest commits as well.

> could you please update the benchmarks in the top-level description to match the current performance?

This is a little bit problematic. The current benchmarks do not correctly represent the overall performance for several reasons:
1. It does not care about the drawing process, we need to simulate the drawing with different brush sizes to correctly measure timings, not just make static calls of methods.
2. The major performance boost from the local updates comes from partial data transfer in VisPy, which is executed asynchronously from the napari code and is hard to measure using the benchmark suite.

I created a benchmark for the caching optimization, but it is not clear for me how to benchmark the local updates. Creating a good benchmark for the whole drawing process is another not very simple task.

**Comment by ksofiyuk** [comment]
> I'm super impressed with the performance you've managed to squeeze out of the CPU-only code

It was an essential thing to do to have good performance :wink:, I played with #3308 a little bit and even with the GPU-based mapping I could hardly see something more than 10-15 FPS with 5000x5000 images.

I think that kind of things should first be optimized for the CPU, it is not a very good practice to cover not asymptotically optimal algorithms just with more powerful hardware like GPUs. I believe that moving computations to the GPU should be performed only as the last resort when all other algorithmic optimizations are exhausted.

**Comment by jni** [comment]
> This is a little bit problematic. The current benchmarks do not correctly represent the overall performance for several reasons:

Ah, I understand; for some reason I thought the benchmarks were new benchmarks measuring partial updates. But,

> I created a benchmark for the caching optimization, but it is not clear for me how to benchmark the local updates. Creating a good benchmark for the whole drawing process is another not very simple task.

Could you not paint a circle or square in the middle of a large canvas using `data_setitem`? ie does a benchmark of `data_setitem` not accurately measure what we are trying to measure here and if not, why not?

**Comment by ksofiyuk** [comment]
> Could you not paint a circle or square in the middle of a large canvas using `data_setitem`? ie does a benchmark of `data_setitem` not accurately measure what we are trying to measure here and if not, why not?

It won't accurately measure the caching optimization as the brush position is static in that situation. The very first call will be very long as it initializes the cache for the whole image. All other calls will have near zero time for mapping as there are no changes.  We need to simulate the moving brush to correctly measure the timings + plus we need to add a parameter that changes the brush size, because with the local updates the size of the brush really matters.

**Comment by ksofiyuk** [comment]
@jni I'm working on a new benchmark for drawing, and I noticed that the benchmarks `Labels2DSuite/Labels3DSuite` can't measure the time for #3308 correctly. The timings in the first message for it are wrong and underestimated for the following reasons:
1. In the benchmark, the Labels layer is created without a viewer, so VispyLabelsLayer is not created, so the mapping code is just not called.
2. Even if you add the VisPy layer, the benchmark cannot measure the mapping time as it is performed in the OpenGL shaders, which are executed asynchronously and do not block the Python code.

So, it would be not that easy to correctly measure the performance of #3308 and we need to be careful when compare it to the CPU-only code in Python.

**Comment by ksofiyuk** [comment]
I added a new benchmark (`benchmark_labels_layer.LabelsDrawing2DSuite`) that measures the timings of brush drawing in different modes (auto/direct, contour == 0/1) with different brush sizes. It simulates the brush drawing from the position (0, 0) to (n - 1, n - 1) with 30 refresh updates along the way.

This PR:
```                                                                                                                          
              ====== ============ ============ ============ ============ ============
              --                                  color_mode / contour
              ------------------- ---------------------------------------------------
                n     brush_size    auto / 0     auto / 1    direct / 0   direct / 1
              ====== ============ ============ ============ ============ ============
               512        8        28.2±0.3ms    33.6±1ms     29.2±1ms    32.9±0.7ms
               512        64       22.8±0.1ms   28.1±0.2ms   22.9±0.1ms   29.1±0.3ms
               512       256       128±0.9ms    153±0.7ms    128±0.7ms     153±1ms
               3072       8        147±0.6ms    165±0.7ms     147±3ms     166±0.4ms
               3072       64       131±0.9ms     151±2ms     132±0.8ms     150±1ms
               3072      256        450±2ms      502±2ms     452±0.7ms     501±2ms
              ====== ============ ============ ============ ============ ============
```

main 7b3f7ecfe0e74af5f52dd3eb380b5752b2f3378c:
```
              ====== ============ ============ ============ ============ ============
              --                                  color_mode / contour
              ------------------- ---------------------------------------------------
                n     brush_size    auto / 0     auto / 1    direct / 0   direct / 1
              ====== ============ ============ ============ ============ ============
               512        8        86.6±0.7ms    242±2ms      100±1ms      256±1ms
               512        64        75.6±1ms     230±1ms      90.6±1ms     241±1ms
               512       256       163±0.9ms     316±2ms      180±1ms      328±1ms
               3072       8        1.20±0.01s   6.33±0.02s   1.80±0.02s   6.90±0.02s
               3072       64       1.15±0.01s    6.30±0s      1.77±0s      6.90±0s
               3072      256       1.30±0.01s   6.40±0.01s    1.92±0s     7.04±0.02s
              ====== ============ ============ ============ ============ ============
```

The difference is most noticeable when the heavy labels mapping code is used (contour=1), with small brush sizes it gives almost 50x performance boost.

Please keep in mind that this benchmark only measures the time it takes to update `layer.data` + the time it takes to convert labels to colors before sending them to VisPy. This benchmark does not take into account the time it takes to transfer data from napari to OpenGL or the time it takes to process labels in any type of OpenGL shader. And the local updates also give a substantial speed up in the part that is not measured.

**Comment by ksofiyuk** [comment]
> Another thing I'd like explicit testing for, now that you are tracking painted pixels, is that labels_layer.data[i, j] += 1; labels_layer.refresh() updates things correctly.

@jni This PR doesn't affect the behavior of `Labels.refresh()` in any sense. If you call it, it will update the whole labels layer in the same way as it was before. As you can see, I added a new `Labels._partial_labels_refresh` method that only replaces the old `Labels.refresh()` calls in `Labels.data_setitem` and in `Labels._draw`, which is used only during drawing. If other parts of code keep using `refresh`, the local updates do not affect them.

**Comment by ksofiyuk** [comment]
@jni It appears that I have addressed all of your concerns:
* I added a new benchmark that measures the painting time by simulating the process of using the brush https://github.com/napari/napari/pull/5732/commits/8217133079fbe30fa8f9ade67b1298a2c7a89c4f
* I added a test that checks for cache invalidation when color_mode is changed, it catches the bug that I encountered earlier. 6e14b7f45d8af69360e43cea0f3b34c2efcfde8a
* I added a test that checks that local updates work correctly when contour > 0 as it is a non-trivial part of the code where one can easily make a mistake.  0eb2aefe8d21b71866392159a20873ad777e2bc8
* I found and fixed some minor issue: `Labels.get_color` used to call `_raw_to_displayed` to get color, which resulted in unnecessary cache invalidation when you change selected_label.

Please let me know if you have any other concerns.

**Review by jni — APPROVED** [review]
❤️ 🚀

**Comment by jni** [comment]
Hmm, weird, my browser added a cached comment from weeks ago to the review! Ignore it!

**Comment by jni** [comment]
Fantastic update of the PR description, too, I think that's ready to be the merge commit message.

@ksofiyuk once we get approval we typically wait about 24h before merging in case anyone has any final comments. But I expect this to be in main very shortly! Thank you for your patience and persistence here! I think it's a really nice implementation in the end!

**Comment by ksofiyuk** [comment]
I've found a bug that was hidden before the change https://github.com/napari/napari/pull/5732/commits/bfa06c756d424dd44064e5b84dc43c0ea40712b5 . The cache also needs to be invalidated when the seed is changed. The color shuffle feature is broken without it. Before https://github.com/napari/napari/pull/5732/commits/bfa06c756d424dd44064e5b84dc43c0ea40712b5  it worked "correctly" because `Labels.get_color` always implicitly invalidated the cache by passing an array of size 1 to `_raw_to_displayed`.

I fixed it and added a test.

**Comment by jni** [comment]
> I've found a bug that was hidden before the change https://github.com/napari/napari/commit/bfa06c756d424dd44064e5b84dc43c0ea40712b5 . The cache also needs to be invalidated when the seed is changed. The color shuffle feature is broken without it. Before https://github.com/napari/napari/commit/bfa06c756d424dd44064e5b84dc43c0ea40712b5 it worked "correctly" because Labels.get_color always implicitly invalidated the cache by passing an array of size 1 to _raw_to_displayed.

I'm just going to reiterate my earlier [comment](https://github.com/napari/napari/pull/5732#issuecomment-1538025149) 😂 

> Cache invalidation is one of the trickiest things to get right, and I definitely don't trust myself to do it in #3308 in the absence of tests here!

Thanks for adding the new test! 🚀

**Comment by kevinyamauchi** [comment]
Really nice work. Thank you @ksofiyuk !

**Pull request merged**
